src/Algebra/Graph/Acyclic/AdjacencyMap.hs

```diff
@@ -0,0 +1,449 @@
+  { aam :: AM.AdjacencyMap a
+  } deriving (Show, Eq, Ord)
+
+-- | Check if the internal graph representation is consistent,
+-- i.e. that all edges refer to existing vertices and the graph
+-- is acyclic. It should be impossible to create an inconsistent 
+-- adjacency map.
+--
+-- @
+-- consistent 'empty'                 == True
+-- consistent (vertex x)            == True
+-- consistent (disjointOverlay x y)        == True
+-- consistent (disjointConnect x y)        == True
+-- consistent (vertices x)          == True
+-- consistent (box x y)             == True
+-- consistent (transitiveClosure x) == True
+-- consistent (transpose x)         == True
+-- consistent (fromGraph (<) x)     == True
+-- consistent (fromGraph (>) x)     == True
+-- consistent (1 + 2)               == True
+-- consistent (1 * 2 + 2 * 3)       == True
+-- @
+consistent :: Ord a => AdjacencyMap a -> Bool
+consistent (AAM m) = AM.consistent m && AM.isAcyclic m
+
+-- | Construct the /empty acyclic graph/.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- isEmpty 'empty'                           == True
+-- isEmpty ('disjointOverlay' 'empty' 'empty')          == True
+-- isEmpty ('vertex' x)                      == False
+-- isEmpty ('removeVertex' x $ 'vertex' x)     == True
+-- isEmpty ('removeEdge' 1 2 $ 1 * 2)        == False
+-- @
+empty :: AdjacencyMap a
+empty = coerce AM.empty
+
+-- | Construct the graph comprising /a single isolated vertex/.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- 'isEmpty'     (vertex x) == False
+-- 'hasVertex' x (vertex x) == True
+-- 'vertexCount' (vertex x) == 1
+-- 'edgeCount'   (vertex x) == 0
+-- @
+vertex :: a -> AdjacencyMap a
+vertex = coerce AM.vertex
+
+-- | Construct the graph comprising a given list of isolated vertices.
+-- Complexity: /O(L * log(L))/ time and /O(L)/ memory, where /L/ is
+-- the length of the given list.
+--
+-- @
+-- vertices []              == 'empty'
+-- vertices [x]             == 'vertex' x
+-- 'hasVertex' x . vertices   == 'elem' x
+-- 'vertexCount' . vertices   == 'length' . 'Data.List.nub'
+-- 'vertexSet'   . vertices   == Set.'Set.fromList'
+-- @
+vertices :: (Ord a) => [a] -> AdjacencyMap a
+vertices = coerce AM.vertices
+
+-- | Perform a disjoint overlay of two different acyclic graphs.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- 'isEmpty' (disjointOverlay x y)             == 'isEmpty' x && 'isEmpty' y
+-- 'hasVertex' (Left z) (disjointOverlay x y)  == 'hasVertex' z x
+-- 'hasVertex' (Right z) (disjointOverlay x y) == 'hasVertex' z y
+-- 'vertexCount' (disjointOverlay x y)         >= 'vertexCount' x
+-- 'vertexCount' (disjointOverlay x y)         == 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (disjointOverlay x y)         >= 'edgeCount' x
+-- 'edgeCount'   (disjointOverlay x y)         == 'edgeCount' x + 'edgeCount' y
+-- 'vertexCount' (disjointOverlay 1 2)         == 2
+-- 'edgeCount'   (disjointOverlay 1 2)         == 0
```

Comment:
These two special cases are also unnecessary -- they are covered by the equality properties above.